### DISCUSSION TITLE
Objective that encourages use of exactly two sinusoids

### DISCUSSION THREAD

**Discussion by mchiquier** [MilesCranmer/PySR#475] [Q&A]
For example, if I wanted to encourage the use of 2 sin functions, can I somehow implement that in Julia? I specifically need the nodes in string form, for down-the-line applications.

As a first step, I tried first converting the Tree into symbols, but I am getting the following error: "JULIA: Please load the `SymbolicUtils` package to use `node_to_symbolic(::Node, ::AbstractOperatorEnum; kws...)`." I tried importing inside the function but that didn't work either.

For example, if I wanted to encourage the use of 2 sin functions, can I somehow implement that in Julia? I specifically need the nodes in string form, for down-the-line applications.

As a first step, I tried first converting the Tree into symbols, but I am getting the following error: "JULIA: Please load the `SymbolicUtils` package to use `node_to_symbolic(::Node, ::AbstractOperatorEnum; kws...)`." I tried importing inside the function but that didn't work either.

```python
model = PySRRegressor(
    niterations=40,  # < Increase me for better results
    binary_operators=["+", "*"],
    unary_operators=[
        "sin",
        "cos",
        "inv(x) = 1/x",
        "tanh",
        "exp"
       # "inv(x) = 1/x",
        # ^ Custom operator (julia syntax)
    ],
    extra_sympy_mappings={"inv": lambda x: 1 / x},
    # ^ Define operator for SymPy as well
    full_objective=""" 
            import Pkg; Pkg.add("SymbolicUtils")
            using SymbolicUtils
            function my_objective(tree::Node{T}, dataset::Dataset{T}, options::Options) where T 
                 prediction, flag = eval_tree_array(tree, dataset.X, options) 
                 print(node_to_symbolic(tree))
                 !flag && return T(Inf) 
                 abs3(x) = abs(x) ^ 3 
                 return sum(abs3, prediction .- dataset.y) / length(prediction) 
            end 
            """,
    )
```

**Comment by MilesCranmer** [comment]
Yeah, you can do this by, e.g., (for `full_objective`)
```julia
function my_objective(tree::Node{T}, dataset::Dataset{T,L}, options::Options) where {T,L}
    prediction, flag = eval_tree_array(tree, dataset.X, options)
    !flag && return convert(L, Inf)

    sin_idx = 1  # Change if you change the order you put `sin`, or use `findfirst(==(sin), options.operators.unaops)::Int`
    
    prediction_loss = sum(i -> abs(prediction[i] - dataset.y[i])^3, eachindex(dataset.y)) / length(dataset.y)


    # Count number of sinusoids:
    num_sins = count(node -> node.degree == 1 && node.op == sin_idx, tree)
    
    # Add penalty of 10 for every sinusoids off from 2:
    regularization = convert(L, 10 * abs(num_sins - 2))
    
    return prediction_loss + regularization
end
```

A lot of nice functions like `count`, `sum`, etc are implemented in https://github.com/SymbolicML/DynamicExpressions.jl/blob/master/src/base.jl, so there's no need to convert to SymbolicUtils.jl.

**Reply by mchiquier** [reply]
Thanks a ton! Actually, is there any way I could get the tree in String format too though? I.e. just get the resulting equation in string format?

**Reply by MilesCranmer** [reply]
Yeah you can call `string_tree(tree, options)`.

**Reply by mchiquier** [reply]
Thank you!!
